feat(alerts):  Adding custom Header and StyledPanel components for use in alerts&reports modal redesign

fisjac opened the pull request:
<!---
Please write the PR title following the conventions at https://www.conventionalcommits.org/en/v1.0.0/
Example:
fix(dashboard): load charts correctly
-->

### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Context: This change is part of a broader design proposal https://github.com/apache/superset/discussions/25729 to change the Alerts & Reports modal requested by the Preset team. 

Added two components to the alerts frontend directory in preparation for future changes to the alerts&reports modal: 
(i) StyledPanel component which adds custom styling to the Antd Collapse.Panel component.
(ii) A ValidatedPanelHeader component which takes the following props {
  title: string;
  subtitle: string;
  required: boolean;
  validateCheckStatus: boolean;
};

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
<!--- Skip this if not applicable -->

#### BEFORE

![image](https://github.com/apache/superset/assets/41238731/ded096c9-8774-464d-830e-884429802b61)

#### AFTER

![image](https://github.com/apache/superset/assets/41238731/70a53379-f6c3-401c-b4fc-e2856a5ac464)
### TESTING INSTRUCTIONS
<!--- Required! What steps can be taken to manually verify the changes? -->

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Required feature flags:
- [x] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

fisjac committed:
creating StyledPanel and CustomValidationHeader components for alerts&reports

hughhhh commented on superset-frontend/src/components/Panel/index.tsx:
```diff
@@ -0,0 +1,106 @@
+        <span className="validation-checkmark">{checkmark}</span>
+      </div>
+      <p className="collapse-panel-subtitle">
+        {/* @ts-ignore */}
```
why do we need the ts-ignore here?

fisjac replied:
Since props.header can possibly take type string, TS gets angry saying that there's no property "title" on type string. Due to the conditional check in lines 65-67, we know that if this logic block is reached, props.header will be an object containing the title property.

hughhhh commented on superset-frontend/src/components/Panel/index.tsx:
```diff
@@ -0,0 +1,106 @@
+  );
+
+  // eslint-disable-next-line no-param-reassign
+  props.header = header;
```
why do we need to reassign the `props.header` here?

hughhhh commented on superset-frontend/src/components/Panel/index.tsx:
```diff
@@ -0,0 +1,106 @@
+  }
+
+  let asterisk;
+  // @ts-ignore
```
why do we need the ts-ignore here?

john-bodley commented:
@fisjac would you mind updating the PR description with a more high level overview/context of what the goal of this change is and how you intent to leverage the enriched metadata? It's not apparent what the current use case is.

michael-s-molina reviewed (changes requested):
Thanks for the PR @fisjac. I can understand the need to customize the header in your installation. I think the best approach here would be to actually receive the Panel component as the parameter instead of creating one that might be sufficient for your use case but not for others. By receiving a generic component, we allow all types of header customization.

michael-s-molina commented:
> @michael-s-molina I've refactored the code such that it is only adding new components to the alerts&reports components where it we be implemented. Changes to the Collapse component have been reverted. @hughhhh In the refactor, I've removed the need for override the ts checks, so all of the `@ts-ignore` lines have been removed.

@fisjac I'm trying to better understand the context for this PR. Are you planning to change the default implementation of Alerts and Reports of Superset or override the default implementation of Superset in your installation? The reason I'm asking is because the components you created are not referenced anywhere.

If you are trying to override the default implementation of Superset in your installation, then these components don't belong to Superset codebase but to your codebase. What the Superset Collapse panel component needs to provide is a property that allow you to override the header component.

fisjac commented:
> > @michael-s-molina I've refactored the code such that it is only adding new components to the alerts&reports components where it we be implemented. Changes to the Collapse component have been reverted. @hughhhh In the refactor, I've removed the need for override the ts checks, so all of the `@ts-ignore` lines have been removed.
> 
> @fisjac I'm trying to better understand the context for this PR. Are you planning to change the default implementation of Alerts and Reports of Superset or override the default implementation of Superset in your installation? The reason I'm asking is because the components you created are not referenced anywhere.
> 
> If you are trying to override the default implementation of Superset in your installation, then these components don't belong to Superset codebase but to your codebase. What the Superset Collapse panel component needs to provide is a property that allow you to override the header component.

The context for this PR is in response to the broader design proposal linked here: https://github.com/apache/superset/discussions/25729 While these components are currently orphaned, the idea is to implement them within the redesign of the Alerts&Reports Modal, which is where I've placed the components within the codebase.

michael-s-molina commented:
> The context for this PR is in response to the broader design proposal linked here: https://github.com/apache/superset/discussions/25729 While these components are currently orphaned, the idea is to implement them within the redesign of the Alerts&Reports Modal, which is where I've placed the components within the codebase.

Oh, now I understood. There's still a long way to go regarding that discussion, which will become a SIP (Superset Improvement Proposal) and be submitted for voting before implementation begins. This means that a lot can change regarding the design. I'll add the `hold` label to this PR and revisit after the SIP is approved.

geido commented:
The work was moved in a feature branch and merged. Closing this one
